### PR TITLE
Round12: model-layers=3 — reduce transformer depth (SOTA=4, never varied)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

SOTA uses `--model-layers 4`. Across all ~60+ tay experiments, the **transformer layer depth has never been varied** — it is a completely untested architectural dimension alongside `--model-heads` (askeladd testing 4 vs SOTA 8). Reducing depth to 3 layers tests whether the current 4-layer stack is over-parameterized for the aerodynamic field prediction task at this training budget.

Two mechanisms could help:
1. **Reduced overfitting:** fewer parameters on a 9-epoch budget → better generalization
2. **Increased throughput:** less compute per epoch → model explores more of the loss landscape within the 270-minute training timeout, potentially completing more high-quality gradient steps in the final epochs

**Single delta from SOTA:** only `--model-layers 3` changes (SOTA=4, unchanged).

## Current Baseline (SOTA — PR #115, thorfinn, W&B run `d03oghpp`)

| Metric | Val | Test |
|---|---|---|
| abupt_axis_mean_rel_l2_pct | **9.484%** | **10.580%** |
| surface_pressure | 6.007% | 6.763% |
| volume_pressure | 5.896% | 12.740% |
| wall_shear | 10.632% | 12.237% |

**Merge bar:** val_abupt < 9.484%

## SOTA Config (all flags identical except `--model-layers`)

```
optimizer: lion
lr: 1e-4
weight-decay: 5e-4
lion-beta1: 0.9
lion-beta2: 0.99
no-compile-model: true
batch-size: 4
validation-every: 1
train-surface-points: 65536
eval-surface-points: 65536
train-volume-points: 65536
eval-volume-points: 65536
model-layers: 4         ← SOTA (YOUR RUN CHANGES THIS TO 3)
model-hidden-dim: 512
model-heads: 8
model-slices: 128
ema-decay: 0.999
lr-cosine-t-max: 50
```

## Train Command

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 3 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 \
  --wandb-group tay-round12-model-layers-3
```

**Critical: Only `--model-layers 3` differs from SOTA. Keep every other flag as listed above.**

## Results (fill in after training)

| Metric | Ep 1 | Ep 3 | Ep 5 | Ep 7 | Best | SOTA |
|---|---|---|---|---|---|---|
| val_abupt | | | | | | 9.484% |
| surface_pressure | | | | | | 6.007% |
| volume_pressure | | | | | | 5.896% |
| wall_shear | | | | | | 10.632% |

W&B run ID: _________  
Best epoch: _________  
Beats SOTA (val_abupt < 9.484%)? YES / NO

## Suggested follow-ups (if results are promising)
- `--model-layers 5` (deeper — test the opposite direction)
- `--model-layers 3 --model-heads 4` (combined with askeladd's result if both improve)

### CODE REVIEW

**Comment by morganmcg1** [comment]
@edward — new experiment assigned! Please run the command in the PR body exactly as written.

**What you're testing:** The transformer layer depth (`--model-layers`) has never been varied across all ~60 tay experiments. SOTA uses 4 layers. You're testing 3 layers — the hypothesis is that shallower depth may reduce overfitting on the limited 9-epoch budget and, as a bonus, reduced compute per epoch may allow the model to complete more quality gradient steps within the 270-minute training timeout.

**Single delta rule:** Only `--model-layers 3` changes from SOTA. Keep every other flag identical.

**Merge bar:** val_abupt < 9.484% (SOTA PR #115).

Your detailed analysis of PR #195 was excellent — that level of mechanistic understanding is exactly what we need. Keep it up for this run too. Report your W&B run ID and per-epoch val_abupt in a comment here when training completes.

**Comment by morganmcg1** [comment]
ADVISOR: Trajectory check (running rank0 x11pua0w, after first batch failed at step=0).

Curve so far:
- ep1=61.961%
- ep2=27.021%
- ep3=18.931%
- ep4=15.551%
- ep5=13.523% (delta=+4.039% vs SOTA)

Tracking as a regression — depth=3 is meaningfully behind the SOTA depth=4 trajectory at the same epoch. Combined with nezuko's slices=64 regression and askeladd's heads=4 regression, the **"shrink the model" axis is closed off**: SOTA's 4 layers / 8 heads / 128 slices is well-tuned. The forward question becomes width / FFN ratio (tanjiro #241 hidden-dim=768, frieren #240 mlp-ratio=8) — let those run.

Allow this to complete for the full curve; will record the closeout when ep9 lands.

**Comment by morganmcg1** [comment]
ADVISOR: Your branch has a merge conflict with the base branch and needs to be rebased before it can be merged. Please rebase your branch onto the latest tay branch and resolve any conflicts, then push your branch to update this PR.

**Comment by morganmcg1** [comment]
STUDENT edward:

## Results

**Outcome: REGRESSION confirmed.** Reducing transformer depth from 4 → 3 layers under-performs SOTA on every primary metric. Throughput hypothesis was *partially* validated — the depth-3 stack ran 12 val epochs in the 270-min budget vs SOTA's 9 epochs — but the extra epochs only narrowed the gap; they did not close it.

### val_primary metrics (W&B run `x11pua0w`, best epoch 12)

| Metric | Best (ep12) | PR #115 SOTA cited in PR | Δ vs cited | Current SOTA (PR #222 fern) | Δ vs current |
|---|---:|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.459%** | 9.484% | **+0.975pp** | 9.291% | +1.168pp |
| `surface_pressure_rel_l2_pct` | 6.735% | 6.007% | +0.728pp | 5.871% | +0.864pp |
| `volume_pressure_rel_l2_pct` | 6.566% | 5.896% | +0.670pp | 5.879% | +0.687pp |
| `wall_shear_rel_l2_pct` | 11.709% | 10.632% | +1.077pp | 10.342% | +1.367pp |

**Beats SOTA (val_abupt < 9.484%)? NO** — regression on every axis.

### Per-epoch validation curve

| ep | val_abupt | surf_p | vol_p | wall_s |
|---:|---:|---:|---:|---:|
|  1 | 61.961 | 47.546 | 42.869 | 67.694 |
|  2 | 27.021 | 19.055 | 16.039 | 30.017 |
|  3 | 18.931 | 12.946 | 10.893 | 21.226 |
|  4 | 15.551 | 10.583 |  8.883 | 17.454 |
|  5 | 13.523 |  9.113 |  7.795 | 15.209 |
|  6 | 12.357 |  8.202 |  7.292 | 13.904 |
|  7 | 11.562 |  7.612 |  6.821 | 13.032 |
|  8 | **69.585** | 61.537 | 22.601 | 81.849 |
|  9 | 14.960 |  9.989 |  9.645 | 16.583 |
| 10 | 11.645 |  7.710 |  7.172 | 12.997 |
| 11 | 10.557 |  6.839 |  6.379 | 11.889 |
| **12** | **10.459** | **6.735** | **6.566** | **11.709** |

### Notable: ep8 instability spike

Val_abupt jumped 11.56 → 69.59 → 14.96 between ep7 → ep8 → ep9 (a +58pp spike on val with no commensurate train spike — train/loss continued descending through this region). Recovery took 2 epochs to surpass the pre-spike best (ep10=11.65 < ep7=11.56). This kind of mid-training val spike has not been seen at depth=4 with the same hyperparams, suggesting the shallower stack has a thinner stability margin under Lion's high-momentum updates near the loss-landscape ridge. **Without ep8 disruption the curve trajectory might have closed the gap by ~0.3-0.5pp**, but even an extrapolated clean curve would not have beaten SOTA.

### Train command (run as written in PR body)

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 3 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 \
  --wandb-name "edward/round12-model-layers-3" \
  --wandb-group tay-round12-model-layers-3
```

### Run details

- **W&B run ID:** `x11pua0w` (rank0). Group: `tay-round12-model-layers-3`
- **Best epoch:** 12 (12 val epochs total, EMA checkpoint)
- **Runtime:** 270.6 min (hit 270-min timeout)
- **Peak GPU memory:** 94.17 GiB (~98.5% utilisation per H100/H200)
- **Total optimizer steps:** 30,446
- **Beats merge bar (val_abupt < 9.484%):** NO

### What happened — analysis

1. **Mechanism 1 (reduced overfitting): NOT supported.** The 3-layer model is *under*-fit, not better-generalising — train loss continues decreasing through ep12 with no sign of overfitting. The depth-4 SOTA already underfits the training set within 9 epochs; removing capacity makes underfit worse, not better. Lower depth = lower expressive power for the multi-physics regression (surface pressure, wall shear vector, volume pressure jointly), and the loss surface at this scale rewards the extra capacity.

2. **Mechanism 2 (more epochs in budget): partially supported.** Depth=3 completed 12 epochs vs SOTA's 9 epochs (+33% more passes). But the marginal-epoch improvement is sub-linear: ep9→ep12 only gained ~0.5pp on val_abupt (14.96 → 10.46, dragged by ep8 spike), and the curve was still descending. With sufficient epochs at infinite budget, depth=3 might close more of the gap, but not all — the asymptotic ceiling appears worse for the shallower stack.

3. **Architectural takeaway (aligned with advisor):** Combined with nezuko's `slices=64` regression (PR not-merged) and askeladd's `heads=4` regression (in-progress), this confirms the **"shrink the model" axis is closed off**. SOTA's `4 layers / 512 hidden / 8 heads / 128 slices` is well-tuned for the 270-min budget; reducing any single axis hurts. The forward question is **width/FFN ratio** (tanjiro #241 hidden-dim=768, frieren #240 mlp-ratio=8) — those tests are the productive next moves.

4. **ep8 spike (new finding):** Lion + depth=3 produced a transient val divergence at ep8 that depth=4 has not exhibited under the same lr/wd/EMA. Suggests the stability margin is thinner with reduced depth — could be relevant if anyone explores depth=3 in combination with higher lr.

### Suggested follow-ups

- **Do NOT pursue depth=3 further.** Closing this off as a regression — the architecture-shrink axis is now fully exhausted (depth ✗, heads ✗, slices ✗).
- **The interesting axis remaining is depth=5** (PR body suggested this) — but combined with the heads/slices/width results, "expand the model" should be tried *only* in directions that increase expressive capacity per token (FFN width, hidden dim, more heads at fixed dim_per_head), not raw depth. Depth=5 would test if the issue is depth-specific or capacity-specific.
- **Skip combined `--model-layers 3 --model-heads 4`** — both are regressions individually, the compound is unlikely to be useful.
- **If anyone explores depth=3 in future:** beware the ep8-style stability spike. May need lower lr or slower EMA to keep depth=3 stable through training.

**Comment by morganmcg1** [comment]
## Results: REGRESSION — Closing

**W&B Run:** `x11pua0w` (group: `tay-round12-model-layers-3`, rank-0)
**State:** running (rt=275min, step=30458)
**Best val_primary/abupt_axis_mean_rel_l2_pct:** 10.4587% (at best_epoch=12)
**vs SOTA (9.291%):** +1.168pp — confirmed regression

### Analysis
model_layers=3 (reducing from default 4) significantly regresses performance. This confirms the model_layers axis:
- model_layers=3 (this run): +1.168pp regression
- model_layers=4 (SOTA baseline): 9.291%

Note: 8 recent re-launch attempts after rebase all failed at ~0.3min (DDP initialization crash). Even if those had succeeded, the trajectory from the main run shows no path to beating SOTA.

Reducing depth hurts — the 4-layer architecture appears well-calibrated for this task. Closing as a dead end.

**Decision: CLOSE**
